### PR TITLE
First try.

### PR DESCRIPTION
**Pull request by joealsubash**
Added classes for a couple of nodes.

### CODE REVIEW

**Comment by jimbrooke** [comment]
Some comments on the class hierarchy.

- I thought we discussed NodeNameBuilder and agreed it was not necessary ?  You should use uhal::Node::getNode() to avoid ever having to construct the absolute register paths : https://ipbus.web.cern.ch/doc/user/html/software/uhalQuickTutorial.html#how-to-traverse-the-address-table-tree

- What is the purpose of class DTP ?  If this is meant to provide functionality that is common to all nodes (like TimingNode in the timing package), then a) it should be called DTPNode (or similar) and b) it should have some common functions.  Not having any functions means it seems to serve no purpose at all.

- Am I right that the intention here is just to include reset functionality ?  Let's try and construct the full list of classes required to model the dtp firmware address table, as well as the basic functions needed to implement the hfButler functionality sent by Tony.  (Let's use https://github.com/DUNE-DAQ/dtpcontrols/wiki to assemble the list)

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1 @@
+/home/subashj/hep_bham/dune/dtp_v1.2_test/test_ipbus_1.10/df-work/src/dtp-controls/.gitignore
```
Comment (roland-sipos) [inline .gitignore]: Custom path in gitignore???

File: include/dtpcontrols/AllNodeNames.hpp [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+ * AllNodeNames is a namespace which includes all
+ * unique ipbus node names within DTP-firmware
+ * which is part of DUNE DAQ.
+ *
+ * This is part of the DUNE DAQ Software Suite, copyright 2022.
+ * Licensing/copyright details are in the COPYING file that you should have
+ * received with this code.
+ */
+
+#ifndef DTP_INCLUDE_DTP_ALLNODENAMES_HPP_
```
Comment (roland-sipos) [inline include/dtpcontrols/AllNodeNames.hpp]: Include guards don't follow DUNE-DAQ conventions.

File: include/dtpcontrols/DTP.hpp [hunk 3]
```diff
@@ -0,0 +1,43 @@
+/**
+ * @file DTP.hpp
+ *
+ * DTP.hpp is a class derived from uhal::Node.
```
Comment (roland-sipos) [inline include/dtpcontrols/DTP.hpp:4]: Wrong docs references.

File: .gitignore (hunk 1 above)
Comment (joealsubash) [reply .gitignore]: Used a symlink was a bad idea. Shall fix this tomorrow.

File: include/dtpcontrols/AllNodeNames.hpp (hunk 2 above)
Comment (joealsubash) [reply include/dtpcontrols/AllNodeNames.hpp]: Ok, shall fix this too.

File: include/dtpcontrols/NodeNameBuilder.hpp [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+ *
+ * This is part of the DUNE DAQ Software Suite, copyright 2022.
+ * Licensing/copyright details are in the COPYING file that you should have
+ * received with this code.
+ */
+
+#ifndef DTP_INCLUDE_DTP_NODENAMEBUILDER_HPP_
+#define DTP_INCLUDE_DTP_NODENAMEBUILDER_HPP_
+
+#include <iostream>
```
Comment (roland-sipos) [inline include/dtpcontrols/NodeNameBuilder.hpp]: include is not needed

File: src/ControlNode.cpp [hunk 5]
```diff
@@ -0,0 +1,52 @@
+/**
+ * @file ControlNode.cpp
+ *
+ * This is part of the DUNE DAQ Software Suite, copyright 2022.
+ * Licensing/copyright details are in the COPYING file that you should have
+ * received with this code.
+ */
+
+#include <exception>
```
Comment (roland-sipos) [inline src/ControlNode.cpp:9]: ?

**Comment by joealsubash** [comment]
> Some comments on the class hierarchy.
> 
>     * I thought we discussed NodeNameBuilder and agreed it was not necessary ?  You should use uhal::Node::getNode() to avoid ever having to construct the absolute register paths : https://ipbus.web.cern.ch/doc/user/html/software/uhalQuickTutorial.html#how-to-traverse-the-address-table-tree
> 
>     * What is the purpose of class DTP ?  If this is meant to provide functionality that is common to all nodes (like TimingNode in the timing package), then a) it should be called DTPNode (or similar) and b) it should have some common functions.  Not having any functions means it seems to serve no purpose at all.
> 
>     * Am I right that the intention here is just to include reset functionality ?  Let's try and construct the full list of classes required to model the dtp firmware address table, as well as the basic functions needed to implement the hfButler functionality sent by Tony.  (Let's use https://github.com/DUNE-DAQ/dtpcontrols/wiki to assemble the list)

1) I had second thoughts about NodeNameBuilder, when using strings node names can be hard to change(I have identified atleast two cases in old dtp-controls). Its easier to change a variable across a whole project.

2) Yes, it is based along the lines of TimingNode and I shall rename it. It has no functions now but it will have in the future.

3) ControlNode and DataReception are the two components with reset registers, so those are prioritised.

File: src/ControlNode.cpp (hunk 5 above)
Comment (joealsubash) [reply src/ControlNode.cpp:9]: fixed bbe29ec.

File: include/dtpcontrols/NodeNameBuilder.hpp (hunk 4 above)
Comment (joealsubash) [reply include/dtpcontrols/NodeNameBuilder.hpp]: fixed bbe29ec.

File: include/dtpcontrols/AllNodeNames.hpp (hunk 2 above)
Comment (joealsubash) [reply include/dtpcontrols/AllNodeNames.hpp]: fixed bbe29ec.

File: .gitignore (hunk 1 above)
Comment (joealsubash) [reply .gitignore]: fixed bbe29ec.

**Comment by jimbrooke** [comment]
please remove NodeNameBuilder and AllNodeNames from the PR

**Comment by joealsubash** [comment]
2b1727d
Removed NodeNameBuilder and AllNodeNames. Added two new node classes.

File: src/DTP.cpp [hunk 6]
```diff
@@ -0,0 +1,19 @@
+/**
```
Comment (jimbrooke) [inline src/DTP.cpp]: filename should be DTPNode.cpp

File: scripts/ResetFW.cpp [hunk 7]
```diff
@@ -0,0 +1,46 @@
+
```
Comment (jimbrooke) [inline scripts/ResetFW.cpp]: Let's put this in test/ResetFW.cpp

File: scripts/ResetFW.cpp [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+#include "ControlNode.hpp"
+#include <iostream>
+
+using namespace uhal;
+
+int main(int argc, char const* argv[]) {
+
+  uhal::setLogLevelTo(uhal::Debug());
+  uhal::ConnectionManager cm("file://" + std::string(argv[1])); //, {"ipbusflx-2.0"})
+  uhal::HwInterface flx = cm.getDevice("U-SIMUDP-JS");
```
Comment (jimbrooke) [inline scripts/ResetFW.cpp]: device should not be hardcoded

File: src/CentralRouterInterfaceNode.cpp [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+
+CentralRouterInterfaceNode::CentralRouterInterfaceNode(const uhal::Node &node)
+    : DTPNode(node) {}
+
+CentralRouterInterfaceNode::~CentralRouterInterfaceNode() {}
+
+void CentralRouterInterfaceNode::EnableRouterInterface(bool dispatch) {
+  try {    
+    getNode("csr.ctrl.en").write(0x1);
+  } catch (...) {
```
Comment (jimbrooke) [inline src/CentralRouterInterfaceNode.cpp]: why do you catch everything here?  i don't think it's required that this function guarantee no throw, so this should not be here (i think).

in fact, try-catch is rarely used in timing sys code.  so why do you try at all ?

File: src/CentralRouterInterfaceNode.cpp [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+    std::cout << "Error enabling central router interface." << "\n";
+  }
+  if(dispatch) {getClient().dispatch();}
+}
+
+void CentralRouterInterfaceNode::DropEmpty(bool dispatch) {
+  try {    
+    getNode("csr.ctrl.drop_empty").write(0x1);
+  } catch (...) {
```
Comment (jimbrooke) [inline src/CentralRouterInterfaceNode.cpp]: same comment re: exceptions.

File: src/ControlNode.cpp [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+//NodeNameBuilder* nnb;
+
+ControlNode::ControlNode(const uhal::Node& node) : DTPNode(node){}
+
+ControlNode::~ControlNode(){}
+
+void ControlNode::Nuke(bool dispatch) {
+  try {    
+    getNode("csr.ctrl.nuke").write(0x1);
+  } catch (...) {
```
Comment (jimbrooke) [inline src/ControlNode.cpp]: catch all

File: src/ControlNode.cpp [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+    std::cout << "Error nuking control block." << "\n";
+  }
+  if(dispatch) {getClient().dispatch();}
+}
+
+void ControlNode::SoftReset(bool dispatch) {
+  try {
+    getNode("csr.ctrl.soft_rst").write(0x1);
+  } catch (...) {
```
Comment (jimbrooke) [inline src/ControlNode.cpp]: catch all

File: src/ControlNode.cpp [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+    std::cout << "Error soft resetting control block." << "\n";
+  }
+  if(dispatch) {getClient().dispatch();}
+}
+
+void ControlNode::MasterReset(bool dispatch) {
+  try {
+    getNode("csr.ctrl.mst_rst").write(0x1);
+  } catch (...) {
```
Comment (jimbrooke) [inline src/ControlNode.cpp]: catch all

File: src/DataReceptionNode.cpp [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+//NodeNameBuilder* nnb;
+
+DataReceptionNode::DataReceptionNode(const uhal::Node& node) : DTPNode(node){}
+
+DataReceptionNode::~DataReceptionNode(){}
+
+void DataReceptionNode::EnableDataReception(bool dispatch) {
+  try {    
+    getNode("csr.ctrl.en").write(0x1);
+  } catch (...) {
```
Comment (jimbrooke) [inline src/DataReceptionNode.cpp]: catch all

File: src/DataReceptionNode.cpp [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+    std::cout << "Error enabling Data Reception block." << "\n";
+  }
+  if(dispatch) {getClient().dispatch();}
+}
+
+void DataReceptionNode::ResetInputWordCounter(bool dispatch) {
+  try {
+    getNode("csr.ctrl.rst_inctr").write(0x1);
+  } catch (...) {
```
Comment (jimbrooke) [inline src/DataReceptionNode.cpp]: catch all

File: src/DataReceptionNode.cpp [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+    std::cout << "Error resetting input word count." << "\n";
+  }
+  if(dispatch) {getClient().dispatch();}  
+}
+
+void DataReceptionNode::ResetOutputWordCounter(bool dispatch) {
+  try {
+    getNode("csr.ctrl.rst_outctr").write(0x1);
+  } catch (...) {
```
Comment (jimbrooke) [inline src/DataReceptionNode.cpp]: catch all

File: src/DataReceptionNode.cpp [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+    std::cout << "Error resetting output word counter." << "\n";
+  }
+}
+
+void DataReceptionNode::ErrorReset(bool dispatch) {
+  try {
+    getNode("csr.ctrl.err_rst").write(0x1);
+  } catch (...) {
```
Comment (jimbrooke) [inline src/DataReceptionNode.cpp]: catch all

File: src/DataReceptionNode.cpp [hunk 18]
```diff
@@ -0,0 +1,131 @@
...
+    std::cout << "Error writing error reset." << "\n";
+  }
+  if(dispatch) {getClient().dispatch();}  
+}
+
+uhal::ValWord<uint32_t> DataReceptionNode::DualPortRamWriteFlag(bool dispatch) {
+  uhal::ValWord<uint32_t> lDPRWF;  
+  try {
+    lDPRWF = getNode("stat.dpr_wen").read();
+  } catch (...) {
```
Comment (jimbrooke) [inline src/DataReceptionNode.cpp]: catch all

File: src/DataReceptionNode.cpp [hunk 19]
```diff
@@ -0,0 +1,131 @@
...
+  }
+  if(dispatch) {getClient().dispatch();}  
+  return lDPRWF;
+  }
+
+uhal::ValWord<uint32_t> DataReceptionNode::StickyErrorBit(bool dispatch) {
+  uhal::ValWord<uint32_t> lSEB;  
+  try {
+    lSEB =  getNode("stat.err").read();
+  } catch (...) {
```
Comment (jimbrooke) [inline src/DataReceptionNode.cpp]: catch all

File: src/DataReceptionNode.cpp [hunk 20]
```diff
@@ -0,0 +1,131 @@
...
+  }
+  if(dispatch) {getClient().dispatch();}  
+  return lSEB;
+}
+
+uhal::ValWord<uint32_t> DataReceptionNode::BackPressureBit(bool dispatch) {
+  uhal::ValWord<uint32_t> lBPB;  
+  try {
+    lBPB = getNode("stat.bp").read();
+  } catch (...) {
```
Comment (jimbrooke) [inline src/DataReceptionNode.cpp]: catch all

File: src/DataReceptionNode.cpp [hunk 21]
```diff
@@ -0,0 +1,131 @@
...
+  }
+  if(dispatch) {getClient().dispatch();}  
+  return lOOS;
+}
+
+uhal::ValWord<uint32_t> DataReceptionNode::RxWordCount(bool dispatch) {
+  uhal::ValWord<uint32_t> lRXWC;  
+  try {
+    lRXWC = getNode("rcvd_wrd_cnt").read();
+  } catch (...) {
```
Comment (jimbrooke) [inline src/DataReceptionNode.cpp]: catch all

File: src/DataReceptionNode.cpp [hunk 22]
```diff
@@ -0,0 +1,131 @@
...
+  }
+  if(dispatch) {getClient().dispatch();}  
+  return lRXWC;
+}
+
+uhal::ValWord<uint32_t> DataReceptionNode::SentWordCount(bool dispatch) {
+  uhal::ValWord<uint32_t> lSWC;
+  try {
+    lSWC =  getNode("sent_wrd_cnt").read();
+  } catch (...) {
```
Comment (jimbrooke) [inline src/DataReceptionNode.cpp]: catch all

File: src/DataReceptionNode.cpp [hunk 23]
```diff
@@ -0,0 +1,131 @@
...
+  }
+  if(dispatch) {getClient().dispatch();}  
+  return lSWC;
+}
+
+uhal::ValWord<uint32_t> DataReceptionNode::SentPacketCount(bool dispatch) {
+  uhal::ValWord<uint32_t> lSPC;
+  try {
+    lSPC = getNode("sent_pkt_cnt").read();
+  } catch (...) {
```
Comment (jimbrooke) [inline src/DataReceptionNode.cpp]: catch all

File: src/FlowMasterNode.cpp [hunk 24]
```diff
@@ -0,0 +1,69 @@
...
+
+FlowMasterNode::~FlowMasterNode(){}
+
+void FlowMasterNode::SourceSelect(const std::string& source, const bool dispatch) {
+  std::map<std::string, std::uint16_t> src_map{{"gbt", 0}, {"wibtor", 1}};
+  auto exists = src_map.find(source);
+  if (exists != src_map.end()) {
+    try {    
+      getNode("csr.ctrl.en").write(src_map[source]);
+    } catch (...) {
```
Comment (jimbrooke) [inline src/FlowMasterNode.cpp]: catch all

File: src/FlowMasterNode.cpp [hunk 25]
```diff
@@ -0,0 +1,69 @@
...
+						{"link1", 1}, {"link2", 1},
+						{"link3", 1}, {"link4", 1}};
+  std::map<std::string, std::uint16_t> sink_link{{"hits", 0}, {"link0", 0},
+						{"link1", 1}, {"link2", 2},
+						{"link3", 3}, {"link4", 4}};
+  auto exist_ss = sink_src.find(sink);
+  if (exist_ss != sink_src.end()) {
+    try {    
+      getNode("csr.ctrl.sink_src").write(sink_src[sink]);
+    } catch (...) {
```
Comment (jimbrooke) [inline src/FlowMasterNode.cpp]: catch all

File: src/FlowMasterNode.cpp [hunk 26]
```diff
@@ -0,0 +1,69 @@
...
+    }
+    if(dispatch) {getClient().dispatch();}
+  }
+  else std::cout<<"Unknown sink source. Please check."<<std::endl;
+
+  auto exists_sl = sink_link.find(sink);
+  if (exists_sl != sink_link.end()) {
+    try {    
+      getNode("csr.ctrl.").write(sink_link[sink]);
+    } catch (...) {
```
Comment (jimbrooke) [inline src/FlowMasterNode.cpp]: catch all

**Comment by jimbrooke** [comment]
Please remove the try-catch blocks.  Read https://dune-daq-sw.readthedocs.io/en/latest/packages/styleguide/#73-exceptions for guidelines on `catch (...)`.  There is only one place it is acceptable, which is not here.  And since you do not catch anything else, you don't need to try.

Please model ALL functions on timing system functions.  We do not have time to debate anything 'extra' on the timescale of the coming release - additions will only slow us down.

**Comment by jimbrooke** [comment]
@joealsubash please can you resolve the comments above?  Please don't pull any new functionality into the PR without discussion.

The only other features I think that would be OK for this PR are to extend the Node classes/methods to cover additional firmware functionality.  Once we have a baseline set of Node classes/functions in develop,  we can do some work in parallel (eg. develop python bindings).

**Comment by joealsubash** [comment]
Thanks for the feedback. Removing all try-catch clauses in next commit. Currently working only on classes(nodes) which apply to the 2.10 release.

**Comment by joealsubash** [comment]
13854d4
Removed try-catch, two new classes added. Removed place-holder functions in CentralRouterInterfaceNode.

File: src/CentralRouterInterfaceNode.cpp (hunk 9 above)
Comment (joealsubash) [reply src/CentralRouterInterfaceNode.cpp]: changed 13854d4

File: src/CentralRouterInterfaceNode.cpp (hunk 10 above)
Comment (joealsubash) [reply src/CentralRouterInterfaceNode.cpp]: changed 13854d4

File: src/ControlNode.cpp (hunk 11 above)
Comment (joealsubash) [reply src/ControlNode.cpp]: changed 13854d4

File: src/ControlNode.cpp (hunk 12 above)
Comment (joealsubash) [reply src/ControlNode.cpp]: changed 13854d4

File: src/ControlNode.cpp (hunk 13 above)
Comment (joealsubash) [reply src/ControlNode.cpp]: changed 13854d4

File: src/DataReceptionNode.cpp (hunk 14 above)
Comment (joealsubash) [reply src/DataReceptionNode.cpp]: changed 13854d4

File: src/DataReceptionNode.cpp (hunk 15 above)
Comment (joealsubash) [reply src/DataReceptionNode.cpp]: changed 13854d4

File: src/DataReceptionNode.cpp (hunk 17 above)
Comment (joealsubash) [reply src/DataReceptionNode.cpp]: changed 13854d4

File: src/DataReceptionNode.cpp (hunk 18 above)
Comment (joealsubash) [reply src/DataReceptionNode.cpp]: changed 13854d4

File: src/DataReceptionNode.cpp (hunk 16 above)
Comment (joealsubash) [reply src/DataReceptionNode.cpp]: changed 13854d4

File: src/DataReceptionNode.cpp (hunk 19 above)
Comment (joealsubash) [reply src/DataReceptionNode.cpp]: changed 13854d4

File: src/DataReceptionNode.cpp (hunk 21 above)
Comment (joealsubash) [reply src/DataReceptionNode.cpp]: changed 13854d4

File: src/DataReceptionNode.cpp (hunk 22 above)
Comment (joealsubash) [reply src/DataReceptionNode.cpp]: changed 13854d4

File: src/FlowMasterNode.cpp (hunk 24 above)
Comment (joealsubash) [reply src/FlowMasterNode.cpp]: changed 13854d4

File: src/FlowMasterNode.cpp (hunk 26 above)
Comment (joealsubash) [reply src/FlowMasterNode.cpp]: changed 13854d4

File: src/DataReceptionNode.cpp (hunk 20 above)
Comment (joealsubash) [reply src/DataReceptionNode.cpp]: changed 13854d4

File: src/DataReceptionNode.cpp (hunk 23 above)
Comment (joealsubash) [reply src/DataReceptionNode.cpp]: changed 13854d4

File: src/FlowMasterNode.cpp (hunk 25 above)
Comment (joealsubash) [reply src/FlowMasterNode.cpp]: changed 13854d4

File: src/DTP.cpp (hunk 6 above)
Comment (joealsubash) [reply src/DTP.cpp]: fixed 58bafe0

File: scripts/ResetFW.cpp (hunk 7 above)
Comment (joealsubash) [reply scripts/ResetFW.cpp]: fixed 58bafe0

**Comment by jimbrooke** [comment]
i think we can close this PR as the relevant stuff is now in other PRs/branches.

**Pull request closed without merging**